Sources/NIOFoundationCompat/Codable+ByteBuffer.swift

```diff
@@ -0,0 +1,136 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2019 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import Foundation
+
+extension ByteBuffer {
+    /// Attempts to decode the `length` bytes from `index` using the `JSONDecoder` `decoder` as `T`.
+    ///
+    /// - parameters:
+    ///    - type: The type type that is attempted to be decoded.
+    ///    - decoder: The `JSONDecoder` that is used for the decoding.
+    ///    - index: The index of the first byte to decode.
+    ///    - length: The number of bytes to decode.
+    /// - returns: The decoded value is successful or `nil` if there are not enough readable bytes available.
+
+    @inlinable
+    public func getJSONDecodable<T: Decodable>(_ type: T.Type,
+                                               decoder: JSONDecoder = JSONDecoder(),
+                                               at index: Int, length: Int) throws -> T? {
```

Comment:
Can we provide a default `type` here and shift it to the last argument? That'd be more inline with the equivalent integer methods.

```swift
public func getJSONDecodable<T: Decodable>(at index: Int, length: Int, using decoder: JSONDecoder(), as type: T.Type = T.self)
```

Reply:
a couple of thoughts here:

- `JSONDecoder` _requires_ you to explicitly state the type and I think that's good because for security reasons, type inference might introduce a very subtle bug if it were to infer a different type
- if we accept that the type should be a required argument, then I believe Swift's naming rules want it to be the first parameter. Because this reads well in English:

> Read a JSON decodable _Sandwich_ at index 7 with length 23.

which translates to swift as `readJSONDecodable(Sandwich.self, at: 7, length: 23)`.

The reverse

`readJSONDecodable(at: 7, length: 23, type: Sandwich.self)`

would read as

> read JSON decodable at index 7 with length 23, type Sandwich

Reply:
> * `JSONDecoder` _requires_ you to explicitly state the type and I think that's good because for security reasons, type inference might introduce a very subtle bug if it were to infer a different type

Ah -- good point!

> * if we accept that the type should be a required argument, then I believe Swift's naming rules want it to be the first parameter. Because this reads well in English:
> 
> > Read a JSON decodable _Sandwich_ at index 7 with length 23.
> 
> which translates to swift as `readJSONDecodable(Sandwich.self, at: 7, length: 23)`.
> 
> The reverse
> 
> `readJSONDecodable(at: 7, length: 23, type: Sandwich.self)`
> 
> would read as
> 
> > read JSON decodable at index 7 with length 23, type Sandwich

Makes sense!